Add support for Shelly battery operated devices

thecode opened the pull request:
Add support for Shelly battery operated devices.
Once devices are configured they will keep their last state when Home Assistant is restarted and will update the state when a new update event arrives from the device.
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [X] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example configuration.yaml

```

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes https://github.com/home-assistant/core/issues/40276
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/16361

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [X] The code change is tested and works locally.
- [X] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [X] The code has been formatted using Black (`black --fast homeassistant tests`)
- [X] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [X] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [X] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

thecode committed:
Shelly sleeping devices support

Wait for CoAP update if device is not ready

thecode committed:
Restore entities upon update from device

thecode committed:
Remove leftover

thecode committed:
Fix multiple sensor platforms

Fix multiple sensor platforms
Cleanup
Update requirements to aioshelly version 0.5.4

thecode committed:
Fix pylint

thecode committed:
Fixed bug with SleepingBinarySensor

Fixed bug with SleepingBinarySensor
Cleanup and better class inheritance

chemelli74 reviewed (approved):
Works perfectly. Great job! :thumbsup:

bieniu reviewed (approved):
I tested the migration of battery powered devices from the current version of config entry and everything works as it should. The initialization of devices that sleep when HA started works fine as well. Great work 👍

thecode committed:
Trigger CI

balloob commented on homeassistant/components/shelly/__init__.py:
```diff
@@ -83,59 +69,111 @@ async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry):
+
+    # Sleeping device, make an empty device
+    if device is None:
+        async with async_timeout.timeout(AIOSHELLY_DEVICE_TIMEOUT_SEC):
```
Doesn't need timeout if it doesn't make any connection.

thecode replied:
Removed

balloob commented on homeassistant/components/shelly/__init__.py:
```diff
@@ -83,59 +69,111 @@ async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry):
 
     coap_context = await get_coap_context(hass)
 
+    # First check if device is not sleeping
```
Why even try connecting? The odds of it being online are so small and it halts setup of Home Assistant for `AIOSHELLY_DEVICE_TIMEOUT_SEC` seconds.

balloob replied:
```python
if entry.data["sleep_period"]:
    device = …
else:
    try:
        …
    except
```

thecode replied:
When first adding a device we reach here from `config_flow` after the user was asked to wake up the device

thecode replied:
This part is refactored, based on `sleep_period`:
= 0, regular device setup
= None or device not in registry, wait for device and finish setup.
else sleeping device setup

balloob commented on homeassistant/components/shelly/__init__.py:
```diff
@@ -83,59 +69,111 @@ async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry):
+            )
 
-    hass.data[DOMAIN][DATA_CONFIG_ENTRY][entry.entry_id] = {}
     coap_wrapper = hass.data[DOMAIN][DATA_CONFIG_ENTRY][entry.entry_id][
```
This name should be `device_wrapper`

thecode replied:
we have a `rest_wrapper` also, do you think it makes more sense to call this `device_wrapper`?

thecode replied:
Changed to `device_wrapper` to be aligned with the name of the class.

balloob commented on homeassistant/components/shelly/__init__.py:
```diff
@@ -83,59 +69,111 @@ async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry):
     return True
 
 
+async def async_migrate_entry(hass: HomeAssistant, entry: ConfigEntry):
```
We shouldn't do this with a migration. We can just do this in async_setup_entry if we see that the key is not there. But I think that we should just skip migration and only say that battery devices are supported after this PR and let people repair their old ones ?

thecode replied:
Since the user needs to wake up the device anyhow first time after this PR, I think your approach is better, we should update the docs that sleeping devices which added before HA with this PR should be removed and added again, what do you think?

thecode replied:
Migration removed, will only add data if it is missing

balloob commented on homeassistant/components/shelly/__init__.py:
```diff
@@ -144,16 +182,47 @@ def __init__(self, hass, entry, device: aioshelly.Device):
+        if not self.device.initialized:
+            return
+
+        # Restored device initialized, update entities
```
If this is for initialization, do it in the constructor? Seems weird to do it inside the generic updatae handler.

thecode replied:
Removed from here will restore inside entity

balloob commented on homeassistant/components/shelly/__init__.py:
```diff
@@ -144,16 +182,47 @@ def __init__(self, hass, entry, device: aioshelly.Device):
+            _LOGGER.debug("Restoring Shelly Device - %s", self.name)
+            for block in self.device.blocks:
+                for sensor_id in block.sensor_ids:
+                    description = self.hass.data[DOMAIN][DATA_SENSORS]["sensor"].get(
```
This seems weird. Why wouldn't we do it the other way around -> when a sensor entity is first initialized, it will calculate this info. That way we don't have to pre-calculate it for everything.

There is now a really weird reliance:

Entity relies on Device Wrapper
Device Wrapper needs entity storage in hass.data

balloob replied:
Move this all to `async_added_to_hass` lifecycle callback for the entity in case the entity belongs to a sleepy device.

thecode replied:
Removed, not using anymore, restoring inside entity and not needing the sensors list

balloob commented on homeassistant/components/shelly/binary_sensor.py:
```diff
@@ -123,3 +137,15 @@ class ShellyRestBinarySensor(ShellyRestAttributeEntity, BinarySensorEntity):
+    def is_on(self):
+        """Return true if sensor state is on."""
+
+        return self.attribute_value in [STATE_ON, 1]
```
Why would `attribute_value` be `STATE_ON` and not just a boolean ?

thecode replied:
When this is restored from `async_get_last_state` the restored state is `on` or `off`

balloob commented on homeassistant/components/shelly/config_flow.py:
```diff
@@ -53,13 +53,15 @@ async def validate_input(hass: core.HomeAssistant, host, data):
     """Handle a config flow for Shelly."""
 
-    VERSION = 1
+    VERSION = 2
```
Migrations are really tough, as it prevents users from downgrading and should be avoided as much as possible. Try to just do it inside `async_setup_entry`.

balloob commented on homeassistant/components/shelly/entity.py:
```diff
@@ -19,6 +27,41 @@ async def async_setup_entry_attribute_entities(
         config_entry.entry_id
     ][COAP]
+
+    if wrapper.restored_device:
```
`async_setup_entry_attribute_entities` should be split in two. This if-statement already makes it have two different implementations anyway.

balloob commented on homeassistant/components/shelly/entity.py:
```diff
@@ -320,3 +361,84 @@ def device_state_attributes(self) -> dict:
+        self.entry = entry
+        self.restored_state = None
+
+        if wrapper.restored_device:
```
Isn't this always a restored device because this is the SleepingBlockAttributeEntity?

thecode replied:
Revised the logic here, removed `restored_device`, logic based on `block is None`

thecode committed:
PR review changes

Apply PR review changes
Updated entities from inside blocks
Do not use ` async_migrate_entry`
Improve migration

thecode committed:
Fix pylint

thecode committed:
Rebase & Fix bug in binary sensor

thecode committed:
Fix unload entry

chemelli74 reviewed (approved):
Latest code works fine here! :thumbsup:

healeydave commented:
I have Shelly DW2 sensors that don't update state in HA, can I test this?

balloob commented on homeassistant/components/shelly/__init__.py:
```diff
@@ -83,77 +68,121 @@ async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry):
+            data["sleep_period"] = get_device_sleep_period(device.settings)
+            data["model"] = device.settings["device"]["type"]
+            hass.config_entries.async_update_entry(entry, data=data)
+            entry.data = data
```
Don't do this. This should already be done inside async_update_entry.
```suggestion
```

balloob commented on homeassistant/components/shelly/__init__.py:
```diff
@@ -83,77 +68,121 @@ async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry):
+            hass.config_entries.async_update_entry(entry, data=data)
+            entry.data = data
+
+        hass.loop.create_task(async_device_setup(hass, entry, device))
```
```suggestion
        hass.async_create_task(async_device_setup(hass, entry, device))
```

thecode committed:
Apply suggestions from code review

Co-authored-by: Paulus Schoutsen <paulus@home-assistant.io>

balloob commented on homeassistant/components/shelly/entity.py:
```diff
@@ -320,3 +375,74 @@ def device_state_attributes(self) -> dict:
+        last_state = await self.async_get_last_state()
+
+        if last_state is not None:
+            self.set_last_state(last_state.state)
```
This can't work. We should translate the `last_state.state` to True/False.

balloob commented on homeassistant/components/shelly/entity.py:
```diff
@@ -320,3 +375,74 @@ def device_state_attributes(self) -> dict:
+        if self.block is not None:
+            return super().attribute_value
+
+        return self.last_state
```
This needs to be a boolean value, which it's not with current `set_last_state` implementation.

balloob commented on homeassistant/components/shelly/entity.py:
```diff
@@ -320,3 +375,74 @@ def device_state_attributes(self) -> dict:
+        if self.block is None and self.wrapper.device.initialized:
+            for block in self.wrapper.device.blocks:
+                for sensor_id in block.sensor_ids:
+                    unique_id = f"{self.wrapper.mac}-{block.description}-{sensor_id}"
+                    if unique_id == self.unique_id:
```
I wonder if we can change this around to allow faster comparison:

```python
if self.block or not self.wrapper.device.initialized:
    return

entity_mac, entity_block, entity_sensor = self.unique_id.split('-')

for block in self.wrapper.device.blocks:
    if block.description != entity_block:
       continue    

    for sensor_id in block.sensor_ids:
        if sensor_id != entity_sensor:
            continue

        self.block = block
        break
```

thecode committed:
Apply suggestions from code review

Revise block attach loop
Revise last state handling
